### PR TITLE
Port Enhancements from EA

### PR DESCRIPTION
**Pull request by dvonthenen**
## Proposed changes

This ports enhancements from the Agent EA work that is happening. Since that work will probably take some time to get to GA, backporting the things I can.

Changes:
- Addresses Issue https://github.com/deepgram/deepgram-python-sdk/issues/457
- Adds a `is_muted` call on the Microphone
- Adds a `wait_for_complete_with_mute` call on the Speaker
- Enables a mode where the Speaker has the capability to automatically mute the Microphone when needed
- Rearranging some code

## Types of changes

What types of changes does your code introduce to the community Python SDK?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update or tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I have lint'ed all of my code using repo standards
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

NA

## Summary by CodeRabbit

- **New Features**
	- Enhanced microphone control in the application, allowing users to mute and manage audio input effectively.
	- Added support for microphone functionality in the AsyncSpeakWSClient and SpeakWSClient classes.
	- Introduced a new speech-to-text functionality in the example application.
	- Simplified audio data handling in example applications for improved user experience.

- **Bug Fixes**
	- Improved logging mechanisms to prevent potential errors related to thread handling and HTTP responses.

- **Documentation**
	- Updated example scripts to reflect changes in audio handling and configuration options.

### CODE REVIEW

File: examples/text-to-speech/websocket/simple/main.py [hunk 1]
```diff
@@ -13,20 +13,23 @@
     SpeakWSOptions,
 )
 
-AUDIO_FILE = "output.wav"
-TTS_TEXT = "Hello, this is a text to speech example using Deepgram. How are you doing today? I am fine thanks for asking."
+TTS_TEXT = "Hello, this is a text to speech example using Deepgram."
+
+global warning_notice
+warning_notice = True
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/simple/main.py:19]: this is just an example

File: examples/text-to-speech/websocket/simple/main.py [hunk 2]
```diff
@@ -35,25 +38,46 @@
...
+            global warning_notice
+            if warning_notice:
+                print("Received binary data")
+                print("You can do something with the binary data here")
+                print("OR")
+                print(
+                    "If you want to simply play the audio, set speaker_playback to true in the options for DeepgramClientOptions"
+                )
+                warning_notice = False
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/simple/main.py:50]: this is just an example

File: examples/text-to-speech/websocket/simple/main.py [hunk 3]
```diff
@@ -69,11 +93,13 @@
 
         # send the text to Deepgram
         dg_connection.send_text(TTS_TEXT)
+
         # if auto_flush_speak_delta is not used, you must flush the connection by calling flush()
         dg_connection.flush()
 
         # Indicate that we've finished
-        time.sleep(7)
+        dg_connection.wait_for_complete()
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/simple/main.py:101]: this is just an example and the intended functionality

File: deepgram/audio/microphone/microphone.py [hunk 4]
```diff
@@ -265,13 +259,43 @@ def finish(self) -> bool:
...
+            if self._is_muted:
+                size = len(input_data)
+                input_data = b"\x00" * size
+
+            self._push_callback(input_data)
+        except Exception as e:
+            self._logger.error("Error while sending: %s", str(e))
+            raise
+
+        return input_data, pyaudio.paContinue
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/microphone/microphone.py:301]: This is intented

File: deepgram/audio/microphone/microphone.py [hunk 5]
```diff
@@ -225,19 +197,42 @@
...
+            self._logger.spam("is_muted: stream is None")
+            self._logger.spam("Microphone.is_muted LEAVE")
+            return False
+
+        val = self._is_muted
+
+        self._logger.spam("is_muted: %s", val)
+        self._logger.spam("Microphone.is_muted LEAVE")
+        return val
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/microphone/microphone.py:235]: this is the intented state and return value for this case

File: deepgram/audio/speaker/speaker.py [hunk 6]
```diff
@@ -66,6 +70,7 @@ def __init__(
         channels: int = CHANNELS,
         last_play_delta_in_ms: int = PLAYBACK_DELTA,
         output_device_index: Optional[int] = None,
+        microphone: Optional[Microphone] = None,
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/speaker/speaker.py:73]: Skipping for now.

File: deepgram/audio/speaker/speaker.py [hunk 7]
```diff
@@ -205,6 +189,18 @@
...
+        This method will mute/unmute a Microphone and block until the speak is done playing sound.
+        """
+        self._logger.debug("Speaker.wait_for_complete ENTER")
+
+        mic.mute()
+        self.wait_for_complete()
+        mic.unmute()
+
+        self._logger.debug("Speaker.wait_for_complete LEAVE")
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/speaker/speaker.py:203]: Good catch!

File: deepgram/clients/speak/v1/websocket/async_client.py [hunk 8]
```diff
@@ -62,9 +63,13 @@ class AsyncSpeakWSClient(
...
     _headers: Optional[Dict] = None
 
+    _speaker_created: bool = False
     _speaker: Optional[Speaker] = None
+    _microphone: Optional[Microphone] = None
 
-    def __init__(self, config: DeepgramClientOptions):
+    def __init__(
+        self, config: DeepgramClientOptions, microphone: Optional[Microphone] = None
+    ):
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/async_client.py:72]: Ignoring for now

File: deepgram/clients/speak/v1/websocket/async_client.py [hunk 9]
```diff
@@ -628,6 +640,10 @@
             if await super().finish() is False:
                 self._logger.error("AsyncListenWebSocketClient.finish failed")
 
+            if self._speaker is not None and self._speaker_created:
+                self._speaker.finish()
+                self._speaker_created = False
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/async_client.py:646]: This is intented.

File: examples/text-to-speech/websocket/output_to_wav/main.py [hunk 10]
```diff
@@ -38,46 +35,25 @@
...
-
-        def on_flush(self, flushed, **kwargs):
-            print(f"\n\n{flushed}\n\n")
-
-        def on_clear(self, clear, **kwargs):
-            print(f"\n\n{clear}\n\n")
+            print("Received binary data")
+            with open(AUDIO_FILE, "ab") as f:
+                f.write(data)
+                f.flush()
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/output_to_wav/main.py:41]: This is just an example and this is the intended functionality

File: examples/text-to-speech/websocket/output_to_wav/main.py [hunk 11]
```diff
@@ -38,46 +35,25 @@
...
-        dg_connection.on(SpeakWebSocketEvents.Warning, on_warning)
-        dg_connection.on(SpeakWebSocketEvents.Unhandled, on_unhandled)
+
+        # Generate a generic WAV container header
+        # since we don't support containerized audio, we need to generate a header
+        header = wave.open(AUDIO_FILE, "wb")
+        header.setnchannels(1)  # Mono audio
+        header.setsampwidth(2)  # 16-bit audio
+        header.setframerate(16000)  # Sample rate of 16000 Hz
+        header.close()
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/output_to_wav/main.py:56]: This is just an example and this is the intended functionality

File: examples/text-to-speech/websocket/output_to_wav/main.py [hunk 12]
```diff
@@ -13,23 +13,20 @@
     SpeakWSOptions,
 )
 
-TTS_TEXT = "Hello, this is a text to speech example using Deepgram."
-
-global warning_notice
-warning_notice = True
+AUDIO_FILE = "output.wav"
+TTS_TEXT = "Hello, this is a text to speech example using Deepgram. How are you doing today? I am fine thanks for asking."
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/output_to_wav/main.py:17]: This is just an example and this is the intended functionality

File: examples/speech-to-text/websocket/replay/main.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+load_dotenv()
+
+
+def main():
+    try:
+        # example of setting up a client config. logging values: WARNING, VERBOSE, DEBUG, SPAM
+        # config: DeepgramClientOptions = DeepgramClientOptions(verbose=verboselogs.DEBUG)
+        # deepgram: DeepgramClient = DeepgramClient("", config)
+        # otherwise, use default config
+        deepgram: DeepgramClient = DeepgramClient()
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/speech-to-text/websocket/replay/main.py:27]: It is using the variable for the API within DeepgramClient.

File: examples/speech-to-text/websocket/replay/main.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        #     print(f"\n\n{speech_started}\n\n")
+
+        # def on_utterance_end(self, utterance_end, **kwargs):
+        #     print(f"\n\n{utterance_end}\n\n")
+
+        def on_close(self, close, **kwargs):
+            print(f"\n\n{close}\n\n")
+
+        def on_error(self, error, **kwargs):
+            print(f"\n\n{error}\n\n")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/speech-to-text/websocket/replay/main.py:54]: This is just an example. It doesn't need to be a production app.

File: examples/speech-to-text/websocket/replay/main.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            language="en-US",
+            encoding="linear16",
+            sample_rate=22050,
+            smart_format=True,
+        )
+
+        print("\n\nPress Enter to stop recording...\n\n")
+        if dg_connection.start(options) is False:
+            print("Failed to start connection")
+            return
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/speech-to-text/websocket/replay/main.py:80]: This is just an example. It doesn't need to be a production app.

File: examples/speech-to-text/websocket/replay/main.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+        # open file for reading
+        with open("microsoft_headquarters.wav", "rb") as f:
+            dg_connection.send(f.read())
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/speech-to-text/websocket/replay/main.py:84]: This is just an example. It doesn't need to be a production app.

File: deepgram/audio/speaker/speaker.py [hunk 17]
```diff
@@ -205,6 +189,20 @@ def start(self, active_loop: Optional[asyncio.AbstractEventLoop] = None) -> bool
...
+        """
+        self._logger.debug("Speaker.wait_for_complete ENTER")
+
+        if self._microphone is not None:
+            mic.mute()
+        self.wait_for_complete()
+        if self._microphone is not None:
+            mic.unmute()
+
+        self._logger.debug("Speaker.wait_for_complete LEAVE")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/speaker/speaker.py:204]: This is intentional. There may be a case when they don't want the Speaker to control the Mic mute/unmute all the time but want a function to mute/unmute the mic for this one instance.

**Review by jpvajda — APPROVED** [review]

**Pull request merged**